File: ml/src/snarl_ml/trajectory.py
"""Trajectory rectification for a per-frame ball track.

The tracker emits one ``(x, y)`` peak or ``None`` per frame; some detections are wrong (a
distractor won the heatmap) or missing (low confidence). A ball follows a smooth path, so the
surrounding detections constrain where it must be. Two stages, mirroring the trajectory
rectification in the TrackNet family this tracker is built on:

1. Gate gross outliers: drop a detection that sits far from the straight line between its nearest
   detected neighbours (a spike onto a distractor), so it cannot corrupt the fit below.
2. Curve fit: over a sliding window, fit a quadratic to the surviving detections, replace any that
   still lie far from it, and fill gaps from it (no extrapolation past the detected span).

Pure and NumPy-only: it operates on the decoded point sequence, independent of the model or
runtime, so it is testable in isolation and portable to the on-device pipeline.
"""

from __future__ import annotations

import numpy as np

from .heatmap import Point

_DEGREE = 2


def _nearest(points: list[Point | None], index: int, step: int) -> tuple[int, Point] | None:
    """Return the nearest non-empty ``(index, point)`` from ``index`` walking by ``step``."""
    j = index + step
    while 0 <= j < len(points):
        point = points[j]
        if point is not None:
            return j, point
        j += step
    return None


def _gate_outliers(points: list[Point | None], max_deviation: float) -> list[Point | None]:
    """Drop detections that deviate from the linear interpolation of their nearest neighbours."""
    gated: list[Point | None] = list(points)
    for i, current in enumerate(points):
        if current is None:
            continue
        previous = _nearest(points, i, -1)
        following = _nearest(points, i, 1)
        if previous is None or following is None:
            continue
        (pi, pp), (fi, fp) = previous, following
        fraction = (i - pi) / (fi - pi)
        expected_x = pp[0] + (fp[0] - pp[0]) * fraction
        expected_y = pp[1] + (fp[1] - pp[1]) * fraction
        if float(np.hypot(current[0] - expected_x, current[1] - expected_y)) > max_deviation:
            gated[i] = None
    return gated


def _fit_eval(
    indices: list[int],
    xs: list[float],
    ys: list[float],
    target: int,
    tolerance: float,
    min_points: int,
) -> Point:
    """Fit a quadratic to the window points (with one outlier-trimming refit) and evaluate it at
    ``target`` frame index."""
    t = np.asarray(indices, dtype=np.float64)
    x = np.asarray(xs, dtype=np.float64)
    y = np.asarray(ys, dtype=np.float64)
    coeff_x = np.polyfit(t, x, _DEGREE)
    coeff_y = np.polyfit(t, y, _DEGREE)
    residual = np.hypot(x - np.polyval(coeff_x, t), y - np.polyval(coeff_y, t))
    inliers = residual <= tolerance
    if not inliers.all() and int(inliers.sum()) >= min_points:
        t, x, y = t[inliers], x[inliers], y[inliers]
        coeff_x = np.polyfit(t, x, _DEGREE)
        coeff_y = np.polyfit(t, y, _DEGREE)
    return round(float(np.polyval(coeff_x, target))), round(float(np.polyval(coeff_y, target)))


def _curve_fill(
    points: list[Point | None],
    window: int,
    curve_tolerance: float,
    min_points: int,
) -> list[Point | None]:
    """Replace off-curve detections and fill gaps from a sliding-window quadratic fit."""
    half = window // 2
    refined: list[Point | None] = list(points)
    for i in range(len(points)):
        lo, hi = max(0, i - half), min(len(points), i + half + 1)
        indices: list[int] = []
        xs: list[float] = []
        ys: list[float] = []
        for j in range(lo, hi):
            nearby = points[j]
            if nearby is not None:
                indices.append(j)
                xs.append(float(nearby[0]))
                ys.append(float(nearby[1]))
        if len(indices) < min_points or i < indices[0] or i > indices[-1]:
            continue
        fit = _fit_eval(indices, xs, ys, i, curve_tolerance, min_points)
        original = points[i]
        if original is None or float(np.hypot(original[0] - fit[0], original[1] - fit[1])) > (
            curve_tolerance
        ):
            refined[i] = fit
    return refined


def rectify_track(
    points: list[Point | None],
    *,
    window: int = 7,
    curve_tolerance: float = 12.0,
    min_points: int = 4,
    max_deviation: float = 25.0,
) -> list[Point | None]:
    """Return a rectified copy of ``points`` with spikes removed, outliers replaced, gaps filled.

    First gross spikes (further than ``max_deviation`` px from the line between their nearest
    detected neighbours) are dropped; then a quadratic is fit over a ``window``-frame span centred
    on each frame, replacing detections more than ``curve_tolerance`` px from it and filling gaps
    that lie within the window's detected span. Frames with fewer than ``min_points`` nearby
    detections are left unchanged. Raises ``ValueError`` on invalid parameters.
    """
    if window < 3:
        raise ValueError(f"window must be at least 3, got {window}")
    if min_points <= _DEGREE:
        raise ValueError(f"min_points must exceed degree {_DEGREE}, got {min_points}")
    if curve_tolerance <= 0.0:
        raise ValueError(f"curve_tolerance must be positive, got {curve_tolerance}")
    if max_deviation <= 0.0:
        raise ValueError(f"max_deviation must be positive, got {max_deviation}")

    gated = _gate_outliers(points, max_deviation)
    return _curve_fill(gated, window, curve_tolerance, min_points)

File: app/src/domain/trajectory.ts
import {Point} from './cricket';

// Trajectory rectification for a per-frame ball track. A ball moves smoothly, so the surrounding
// detections constrain where it must be: gate spikes that jump away from the neighbour line, then
// fit a sliding-window quadratic to replace off-curve detections and fill short gaps. This mirrors
// the rectification step in the TrackNet family the tracker is built on, and is validated against
// ground truth in the ml package (snarl_ml.trajectory) before being relied on here.

const DEGREE = 2;

export interface RectifyOptions {
  readonly window?: number;
  readonly curveTolerance?: number;
  readonly minPoints?: number;
  readonly maxDeviation?: number;
}

interface Quadratic {
  readonly a: number;
  readonly b: number;
  readonly c: number;
}

function evalQuadratic(q: Quadratic, t: number): number {
  return q.a * t * t + q.b * t + q.c;
}

// Least-squares quadratic fit v(t) = a*t^2 + b*t + c via the 3x3 normal equations, solved by
// Cramer's rule. Returns null only when the system is singular (degenerate t values).
function fitQuadratic(ts: readonly number[], vs: readonly number[]): Quadratic | null {
  let s0 = 0;
  let s1 = 0;
  let s2 = 0;
  let s3 = 0;
  let s4 = 0;
  let r0 = 0;
  let r1 = 0;
  let r2 = 0;
  for (let i = 0; i < ts.length; i += 1) {
    const t = ts[i] ?? 0;
    const v = vs[i] ?? 0;
    const tt = t * t;
    s0 += 1;
    s1 += t;
    s2 += tt;
    s3 += tt * t;
    s4 += tt * tt;
    r0 += tt * v;
    r1 += t * v;
    r2 += v;
  }
  // [[s4 s3 s2],[s3 s2 s1],[s2 s1 s0]] · [a,b,c] = [r0,r1,r2]
  const det =
    s4 * (s2 * s0 - s1 * s1) - s3 * (s3 * s0 - s1 * s2) + s2 * (s3 * s1 - s2 * s2);
  if (Math.abs(det) < 1e-9) {
    return null;
  }
  const a = (r0 * (s2 * s0 - s1 * s1) - s3 * (r1 * s0 - s1 * r2) + s2 * (r1 * s1 - s2 * r2)) / det;
  const b = (s4 * (r1 * s0 - s1 * r2) - r0 * (s3 * s0 - s1 * s2) + s2 * (s3 * r2 - r1 * s2)) / det;
  const c = (s4 * (s2 * r2 - r1 * s1) - s3 * (s3 * r2 - r1 * s2) + r0 * (s3 * s1 - s2 * s2)) / det;
  return {a, b, c};
}

function nearest(
  points: readonly (Point | null)[],
  index: number,
  step: number,
): {readonly index: number; readonly point: Point} | null {
  let j = index + step;
  while (j >= 0 && j < points.length) {
    const point = points[j];
    if (point !== null && point !== undefined) {
      return {index: j, point};
    }
    j += step;
  }
  return null;
}

// Drop detections that deviate from the linear interpolation of their nearest detected neighbours.
function gateOutliers(points: readonly (Point | null)[], maxDeviation: number): (Point | null)[] {
  const gated: (Point | null)[] = [...points];
  for (let i = 0; i < points.length; i += 1) {
    const current = points[i];
    if (current === null || current === undefined) {
      continue;
    }
    const previous = nearest(points, i, -1);
    const following = nearest(points, i, 1);
    if (previous === null || following === null) {
      continue;
    }
    const fraction = (i - previous.index) / (following.index - previous.index);
    const expectedX = previous.point.x + (following.point.x - previous.point.x) * fraction;
    const expectedY = previous.point.y + (following.point.y - previous.point.y) * fraction;
    if (Math.hypot(current.x - expectedX, current.y - expectedY) > maxDeviation) {
      gated[i] = null;
    }
  }
  return gated;
}

// Fit a quadratic to the window's detections (with one outlier-trimming refit) and evaluate at the
// target frame index. Returns null only on a degenerate fit.
function fitEval(
  indices: readonly number[],
  xs: readonly number[],
  ys: readonly number[],
  target: number,
  tolerance: number,
  minPoints: number,
): Point | null {
  // Fit in indices centred on the window mean: Σt⁴ grows with the raw frame index (t⁴ ≈ 10⁸ at
  // frame 100), so centring keeps the normal-equations system well-conditioned in float64.
  const offset = indices.reduce((sum, index) => sum + index, 0) / indices.length;
  const centred = indices.map(index => index - offset);
  let fitX = fitQuadratic(centred, xs);
  let fitY = fitQuadratic(centred, ys);
  if (fitX === null || fitY === null) {
    return null;
  }
  const keepT: number[] = [];
  const keepXs: number[] = [];
  const keepYs: number[] = [];
  let trimmed = false;
  for (let i = 0; i < centred.length; i += 1) {
    const t = centred[i] ?? 0;
    const x = xs[i] ?? 0;
    const y = ys[i] ?? 0;
    if (Math.hypot(x - evalQuadratic(fitX, t), y - evalQuadratic(fitY, t)) <= tolerance) {
      keepT.push(t);
      keepXs.push(x);
      keepYs.push(y);
    } else {
      trimmed = true;
    }
  }
  if (trimmed && keepT.length >= minPoints) {
    const refitX = fitQuadratic(keepT, keepXs);
    const refitY = fitQuadratic(keepT, keepYs);
    if (refitX !== null && refitY !== null) {
      fitX = refitX;
      fitY = refitY;
    }
  }
  const centredTarget = target - offset;
  return {
    x: Math.round(evalQuadratic(fitX, centredTarget)),
    y: Math.round(evalQuadratic(fitY, centredTarget)),
  };
}

function curveFill(
  points: readonly (Point | null)[],
  window: number,
  curveTolerance: number,
  minPoints: number,
): (Point | null)[] {
  const half = Math.floor(window / 2);
  const refined: (Point | null)[] = [...points];
  for (let i = 0; i < points.length; i += 1) {
    const lo = Math.max(0, i - half);
    const hi = Math.min(points.length, i + half + 1);
    const indices: number[] = [];
    const xs: number[] = [];
    const ys: number[] = [];
    for (let j = lo; j < hi; j += 1) {
      const nearby = points[j];
      if (nearby !== null && nearby !== undefined) {
        indices.push(j);
        xs.push(nearby.x);
        ys.push(nearby.y);
      }
    }
    const first = indices[0];
    const last = indices[indices.length - 1];
    if (
      indices.length < minPoints ||
      first === undefined ||
      last === undefined ||
      i < first ||
      i > last
    ) {
      continue;
    }
    const fit = fitEval(indices, xs, ys, i, curveTolerance, minPoints);
    if (fit === null) {
      continue;
    }
    const original = points[i];
    if (
      original === null ||
      original === undefined ||
      Math.hypot(original.x - fit.x, original.y - fit.y) > curveTolerance
    ) {
      refined[i] = fit;
    }
  }
  return refined;
}

// Return a rectified copy of the track: spikes removed, off-curve detections replaced, short gaps
// filled. Defaults match the ml-side measured best (window 7, curve 12 px, deviation 25 px).
export function rectifyTrack(
  points: readonly (Point | null)[],
  options: RectifyOptions = {},
): (Point | null)[] {
  const window = options.window ?? 7;
  const curveTolerance = options.curveTolerance ?? 12;
  const minPoints = options.minPoints ?? 4;
  const maxDeviation = options.maxDeviation ?? 25;
  if (window < 3) {
    throw new Error(`window must be at least 3, got ${window}`);
  }
  if (minPoints <= DEGREE) {
    throw new Error(`minPoints must exceed degree ${DEGREE}, got ${minPoints}`);
  }
  if (curveTolerance <= 0) {
    throw new Error(`curveTolerance must be positive, got ${curveTolerance}`);
  }
  if (maxDeviation <= 0) {
    throw new Error(`maxDeviation must be positive, got ${maxDeviation}`);
  }
  const gated = gateOutliers(points, maxDeviation);
  return curveFill(gated, window, curveTolerance, minPoints);
}
